graphula-core/src/Graphula.hs

```diff
@@ -193,14 +201,36 @@ instance (MonadIO m, Applicative n, MonadIO n) => MonadGraphulaFrontend (Graphul
         Just key -> do
           existingKey <- get key
           whenNothing existingKey $ do
-            existingUnique <- checkUnique n
+            let updated = embedKey key n
+            existingUnique <- checkUnique updated
             whenNothing existingUnique $ do
-              insertKey key n
+              insertKey key updated
               getEntity key
   remove key = do
     RunDB runDB <- ask
     lift . runDB $ delete key
 
+-- | Embed generated key components in record
+--
+-- @'insertKey'@ doesn't do this for us, and we want to update the
+-- record before doing a @'checkUnique'@ anyway.
+--
+embedKey :: PersistEntity a => Key a -> a -> a
+embedKey key n = if isNothing $ entityPrimary def
+  then n
+  else fromRight n $ fromPersistValues $ zipWith
+    replaceField
+    (fieldHaskell <$> entityFields def)
+    (toPersistValue <$> toPersistFields n)
+ where
+  def = entityDef $ Just n
```

Comment:
Is there a clearer name than `def`? Is this a `definition`? `def` generally means `default` in Haskell land.

Reply:
Maybe `entDef`? These are the distinct names I see used in `persistent`:

```plaintext
let t = entityDef $ dummyFromUnique u
let entDef = entityDef $ Just record
let entity = entityDef $ Just record
let entityDef' = entityDef records
let ent = entityDef $ Just record
```